## 🧠 Problems with Batch Gradient Descent

Batch Gradient Descent (BGD) is conceptually simple and mathematically accurate,  
but it has several **practical limitations** when dealing with large datasets or real-time learning.

---

### ⚙️ 1. High Computational Cost
- In every epoch, BGD calculates the gradient using **all training samples**.
- This requires performing a full pass through the entire dataset before a single update.
- As the dataset grows, the time to compute each update increases drastically.

🧮 Example:
If the dataset has 1 million rows,  
one gradient update means processing all 1 million samples → **very slow convergence**.

---

### 💾 2. High Memory Usage
- Since all data points are required at once,  
  the entire dataset must fit into **memory (RAM)** during training.
- For very large datasets, this becomes infeasible.

⚠️ Real-world datasets (e.g., terabytes of data) cannot be loaded into memory completely.

---

### 🐢 3. Slow Convergence
- BGD updates weights **only once per epoch**,  
  meaning it waits until all samples are processed before improving the model.
- This leads to **delayed learning** — the model doesn’t start improving until the full dataset is seen.

---

### 🧱 4. Stuck in Local Minima (for Non-Convex Problems)
- Because it uses the **exact average gradient**,  
  the optimization path is very smooth and deterministic.
- This can cause the model to get stuck in **local minima** or flat regions of the loss surface.

---

### ⚡ 5. Not Suitable for Online / Streaming Data
- In real-time applications (like stock prediction, sensor data, or live user inputs),  
  data arrives continuously.
- BGD cannot handle such situations because it requires **all data upfront**  
  before starting any update.

---

### 🚫 Summary

| Problem | Description |
|----------|--------------|
| **Computation** | Full dataset processed per update → very slow |
| **Memory** | Entire data must fit into memory |
| **Convergence** | Weights updated only once per epoch |
| **Local Minima** | Deterministic updates can get stuck |
| **Online Learning** | Cannot adapt to streaming or incremental data |

---

### ✅ Therefore:
> While Batch Gradient Descent is accurate for small datasets,  
> it becomes inefficient and impractical for large-scale or real-time learning.  
> This is why **Stochastic Gradient Descent (SGD)** and **Mini-Batch GD** are preferred in practice.


---
## ⚡ Stochastic Gradient Descent (SGD)

### 🧩 Overview
**Stochastic Gradient Descent (SGD)** is an optimization algorithm used to minimize a loss function and find the best model parameters.  
Unlike **Batch Gradient Descent**, which updates parameters after processing the entire dataset,  
SGD updates the parameters **after every single training sample**.  

This makes it **faster** and **more memory-efficient**, especially for large datasets.

---

### 🧠 How It Works (for Linear Regression)

Prediction for a single sample:

$$
\hat{y}_i = w x_i + b
$$

Error for that sample:

$$
e_i = y_i - \hat{y}_i
$$

Loss for a single sample (Mean Squared Error):

$$
L_i = (y_i - \hat{y}_i)^2
$$

Gradients with respect to parameters:

$$
\frac{\partial L_i}{\partial w} = -2\, e_i\, x_i,
\quad
\frac{\partial L_i}{\partial b} = -2\, e_i
$$

Parameter update rules (using learning rate \( \eta \)):

$$
w \leftarrow w - \eta \frac{\partial L_i}{\partial w}
\quad \Rightarrow \quad
w = w + 2\eta\, e_i\, x_i
$$

$$
b \leftarrow b - \eta \frac{\partial L_i}{\partial b}
\quad \Rightarrow \quad
b = b + 2\eta\, e_i
$$

For multiple features (vector form):

$$
\hat{y}_i = \mathbf{w}^\top \mathbf{x}_i + b
$$

$$
\nabla_{\mathbf{w}} L_i = -2\, e_i\, \mathbf{x}_i,
\quad
\mathbf{w} \leftarrow \mathbf{w} + 2\eta\, e_i\, \mathbf{x}_i,
\quad
b \leftarrow b + 2\eta\, e_i
$$

---

### 🔁 Training Process (Conceptually)

1. Initialize parameters \( w, b \)
2. For each epoch:
   - Randomly select one sample \( (x_i, y_i) \)
   - Predict the output \( \hat{y}_i \)
   - Compute the error \( e_i = y_i - \hat{y}_i \)
   - Update \( w \) and \( b \) using the formulas above
3. Repeat for all samples across multiple epochs until the loss converges

---

### ⚙️ Characteristics

| Feature | Description |
|----------|-------------|
| **Update Frequency** | After every single sample |
| **Gradient Accuracy** | Noisy / Approximate |
| **Speed** | Very fast |
| **Memory Requirement** | Very low |
| **Convergence Behavior** | Fast but oscillatory |
| **Local Minima Handling** | Can escape due to randomness |

---

### ✅ Advantages

1. **Faster learning** for large datasets  
2. **Online learning** — can update continuously as new data arrives  
3. **Memory efficient** — no need to load all data into memory  
4. **Escapes local minima** more easily due to random updates  

---

### ⚠️ Limitations

1. The convergence path is **noisy and unstable**  
2. **Highly sensitive** to the learning rate  
3. May **overshoot** near the global minimum if learning rate is too large  

---

### 🚀 Summary

| Aspect | Batch Gradient Descent | Stochastic Gradient Descent |
|---------|------------------------|------------------------------|
| **Update per** | Full dataset | Single sample |
| **Speed** | Slow | Fast |
| **Memory Use** | High | Low |
| **Convergence Path** | Smooth | Noisy |
| **Large Dataset Handling** | Difficult | Excellent |
| **Online Learning** | Not supported | Supported |

---

> ✅ **In short:**  
> Stochastic Gradient Descent updates the model parameters after every individual sample,  
> which leads to faster learning and better scalability — especially useful for very large or streaming datasets.


In [33]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_squared_error

In [2]:
data = pd.read_csv("Student_Performance.csv")

In [3]:
data.head()

Hours Studied  Previous Scores Extracurricular Activities  Sleep Hours  \
0              7               99                        Yes            9   
1              4               82                         No            4   
2              8               51                        Yes            7   
3              5               52                        Yes            5   
4              7               75                         No            8   

   Sample Question Papers Practiced  Performance Index  
0                                 1               91.0  
1                                 2               65.0  
2                                 2               45.0  
3                                 2               36.0  
4                                 5               66.0

In [5]:
data['Extracurricular Activities'] = data['Extracurricular Activities'].apply(lambda x: 1 if x == 'Yes' else 0)

In [8]:
X = data.iloc[:, 0:5].values
Y = data['Performance Index'].values

In [11]:
x_train, x_test, y_train, y_test = train_test_split(X, Y, train_size=0.8)

In [27]:
class MultipleStochastic_GradientDescent:
    def __init__(self):
        self.coef = 0
        self.intercept = 0
    def fit(self, x_train, y_train, learning_rate = 0.001, epoches = 50):
        self.coef = np.ones(x_train.shape[1])
        for i in range(epoches):
            for j in range(x_train.shape[0]):
                row_num = np.random.randint(0, x_train.shape[0])
                y_pred = np.dot(self.coef, x_train[row_num]) + self.intercept
                
                
                new_intercept = (-2 / len(x_train)) * (y_train[row_num] - y_pred)
                self.intercept = self.intercept - (learning_rate * new_intercept)
                
                new_coef = (-2 / len(x_train)) * np.dot((y_train[row_num] - y_pred), x_train[row_num])
                self.coef = self.coef - (learning_rate * new_coef)
                
    def predict(self, x_test):
        return np.dot(x_test, self.coef) + self.intercept

In [28]:
msg = MultipleStochastic_GradientDescent()
msg.fit(x_train, y_train)


In [30]:
print(r2_score(y_test, msg.predict(x_test)))
print(mean_squared_error(y_test, msg.predict(x_test)))

0.8621164045150476
51.28872892028614


## 📊 Model Evaluation and Observation

After training the model using **Stochastic Gradient Descent**,  
the evaluation metrics are as follows:

| Metric | Value |
|---------|--------|
| R² Score | 0.8621164045150476|
| MSE | 51.28872892028614 |

---

### 🔹 Interpretation

The **R² score (0.86)** and **MSE (51.28)** indicate that the model  
is not fitting the data very accurately yet.

However, in this notebook, the **primary objective** was not to  
achieve the best accuracy but to **implement the Stochastic Gradient Descent algorithm manually**  
and understand how it works mathematically and programmatically.

---

### 🔹 Why It’s Not Improved Further

The model’s performance could be improved by:
- Adjusting the **learning rate**
- Increasing the **number of epochs**
- Applying **feature scaling / normalization**
- Or using a **smaller step size**

But these steps are intentionally **not applied here**,  
because the main focus was to demonstrate the **algorithmic logic**  
rather than optimization or model tuning.

---

✅ **In short:**
> This implementation focuses on understanding **how Stochastic Descent works**,  
> not on achieving high accuracy or the best-fit line.
